students/timm67/session04/trigrams.py

```diff
@@ -1,22 +1,179 @@
 #!/usr/bin/env python3
 
-words = "I wish I may I wish I might".split()
+import random
+import sys
 
-def build_trigrams(words):
-    """
-    build up the trigrams dict from the list of words
+trigrams = {}
 
-    returns a dict with:
-       keys: word pairs
-       values: list of followers
-    """
-    trigrams = {}
 
-    # build up the dict here!
+def build_trigrams(word_list, trigrams_dict):
+    if word_list is None:
+        print('build_trigrams: No words to build trigram dict')
+        return
 
-    return trigrams
+    if (len(word_list) < 3):
+        print('build_trigrams: Not enough words to build trigram')
+        return
+
+    for i in range(0, len(word_list)-2, 1):
+        key = (word_list[i], word_list[i+1])
+        if key in trigrams_dict.keys():
+            trigrams_dict[key].append(word_list[i+2])
+        else:
+            trigrams_dict[key] = [word_list[i+2]]
+
+
+def build_trigrams_list(words_list, trigrams_dict):
+    for line_list in words_list:
+        build_trigrams(line_list, trigrams_dict)
+
+
+def generate_text_trigrams(trigram_dict):
+    random_str = str()
+    return_str = str()
+    line_length = 0
+    trigram_keys = list(trigram_dict.keys())
+
+    for i in range(len(trigram_dict)):
+        first_key = random.choice(trigram_keys)
+        print('[{0}] random key: {1}'.format(i, first_key))
+        random_str += (first_key[0])
+        random_str += ' '
+        random_str += (first_key[1])
+        random_str += ' '
+        word_values = trigram_dict[first_key]
+        random_word = random.choice(word_values)
+        print('[{0}] random word value: {1}'.format(i, random_word))
+        random_str += random_word
+        line_length += len(random_str)
+        if (line_length > 50):
+            random_str += '\n'
+            return_str += random_str
+            line_length = 0
+            random_str = ''
+        else:
+            random_str += ' '
+    return return_str
+
+
+def filter_trigram(input_str, exceptions):
+    return ''.join(c for c in input_str if (c.isalpha() or (c in exceptions)))
+
+
+def parse_input_file(in_filename):
+    line_list = []
+    word_list = []
+    exceptions = [' ', '-']
+
+    try:
+        with open(in_filename, 'r') as fd:
+            try:
+                line = fd.readline()
+            except IOError:
+                print("I/O Error with [{0}] readline".format(in_filename))
+                return None
+            while(line):
+                line = line.replace('--', ' ')
+                line = filter_trigram(line, exceptions)
+                if(len(line) == 0):
+                    try:
+                        line = fd.readline()
+                    except IOError:
+                        print("I/O Error with [{0}] readline".format(in_filename))
+                    continue
```

Comment:
Minor: This `if` block seems unnecessary...? If `len(line) == 0`, then it's EOF (end-of-file), and no more `.readline()` will be possible (maybe I'm wrong). No matter what, it'll just print the error message and continue anyway, so doesn't seem necessary.